File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/taskExtensions/backgroundTaskQueue.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
import { default as PQueue } from 'p-queue';

export type WorkItemFunc = () => Promise<void>;

export interface IBackgroundTaskQueue {
    queueBackgroundWorkItem(workItem: WorkItemFunc): void;
}

export class BackgroundTaskQueue implements IBackgroundTaskQueue {
    private readonly queueExecutor: PQueue;

    constructor() {
        this.queueExecutor = new PQueue({
            concurrency: 1,
            autoStart: true
        });
    }

    public queueBackgroundWorkItem(workItem: WorkItemFunc): void {
        if (!workItem) { throw new Error('Missing parameter.  workItem is required'); }

        this.queueExecutor.add(workItem);
    }
}

File: sdk/csharp/libraries/microsoft.bot.solutions/TaskExtensions/BackgroundTaskQueue.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Concurrent;
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.Bot.Solutions.TaskExtensions
{
    public class BackgroundTaskQueue : IBackgroundTaskQueue
    {
        private ConcurrentQueue<Func<CancellationToken, Task>> _workItems = new ConcurrentQueue<Func<CancellationToken, Task>>();
        private SemaphoreSlim _signal = new SemaphoreSlim(0);

        public void QueueBackgroundWorkItem(Func<CancellationToken, Task> workItem)
        {
            if (workItem == null)
            {
                throw new ArgumentNullException(nameof(workItem));
            }

            _workItems.Enqueue(workItem);
            _signal.Release();
        }

        public async Task<Func<CancellationToken, Task>> DequeueAsync(CancellationToken token)
        {
            await _signal.WaitAsync(token).ConfigureAwait(false);

            _workItems.TryDequeue(out var workItem);

            return workItem;
        }
    }
}